File: c/tree/construct-binary-tree-from-preorder-and-inorder-traversal.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *rescan(int *preorder, int left_pre, int *inorder, int left_in, int len)
{
    if(len <= 0)
        return NULL;

    struct TreeNode *node = malloc(sizeof(struct TreeNode));
    int node_pos = left_in;
    while(node_pos < left_in + len){
        if(inorder[node_pos] == preorder[left_pre])
            break;
        ++node_pos;
    }

    node->val = preorder[left_pre];
    node->left = rescan(preorder, left_pre + 1, inorder, left_in, node_pos - left_in);
    node->right = rescan(preorder, left_pre + 1 + node_pos - left_in, inorder,
            node_pos + 1, len - 1 - (node_pos - left_in));

    return node;
}

struct TreeNode* buildTree(int* preorder, int preorderSize, int* inorder, int inorderSize) {
    return rescan(preorder, 0, inorder, 0, inorderSize);
}

File: python/tree/construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rescan(self, left_pre, left_in, count):
        if count <= 0:
            return None

        node = TreeNode(self.preorder[left_pre])
        node_pos = self.inorder.index(self.preorder[left_pre])

        node.left = self.rescan(left_pre + 1, left_in, node_pos - left_in)
        node.right = self.rescan(left_pre + 1 + (node_pos - left_in), \
            node_pos + 1, count - 1 - (node_pos - left_in))
        return node

    def buildTree(self, preorder, inorder):
        self.preorder = preorder
        self.inorder = inorder

        return self.rescan(0, 0, len(inorder))
